pkg/controller/v1alpha2/applicationconfiguration/component.go

```diff
@@ -169,9 +179,74 @@ func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtim
 		return false
 	}
 	c.Logger.Info(fmt.Sprintf("ControllerRevision %s created", revisionName))
+	if int64(c.RevisionLimit) < nextRevision {
+		if err := c.cleanupControllerRevision(curComp); err != nil {
+			c.Logger.Info(fmt.Sprintf("failed to clean up revisions of Component %v.", err))
+		}
+	}
 	return true
 }
 
+// clean revisions when over limits
+func (c *ComponentHandler) cleanupControllerRevision(curComp *v1alpha2.Component) error {
+	labels := &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			ControllerRevisionComponentLabel: curComp.Name,
+		},
+	}
+	selector, err := metav1.LabelSelectorAsSelector(labels)
+	if err != nil {
+		return err
+	}
+
+	// List and Get Object, controller-runtime will create Informer cache
+	// and will get objects from cache
+	revisions := &appsv1.ControllerRevisionList{}
+	if err := c.Client.List(context.TODO(), revisions, &client.ListOptions{LabelSelector: selector}); err != nil {
+		return err
+	}
+
+	// Get appConfigs and workloads filter controllerRevision used
+	appConfigs := &v1alpha2.ApplicationConfigurationList{}
+	if err := c.Client.List(context.Background(), appConfigs); err != nil {
+		return err
+	}
+
+	// get all revisions used and skipped
+	liveHashes := make(map[string]bool)
+	for _, appConfig := range appConfigs.Items {
+		for _, component := range appConfig.Spec.Components {
+			if component.RevisionName != "" {
+				liveHashes[component.RevisionName] = true
+			}
+		}
+	}
+
+	toKeep := c.RevisionLimit + len(liveHashes)
+	toKill := len(revisions.Items) - toKeep
+	if toKill <= 0 {
+		return nil
+	}
+	// Clean up old revisions from smallest to highest revision (from oldest to newest)
+	sort.Sort(historiesByRevision(revisions.Items))
+	for _, revision := range revisions.Items {
+		if toKill <= 0 {
+			break
+		}
+		if hash := revision.GetName(); liveHashes[hash] {
+			continue
+		}
+		// Clean up
+		revisionToClean := revision
+		if err := c.Client.Delete(context.TODO(), &revisionToClean); err != nil {
+			return err
+		}
+		c.Logger.Info(fmt.Sprintf("ControllerRevision %s deleted", revision.Name))
+		toKill--
+	}
+	return nil
```

Comment:
Could you split it into two parts(functions), one part is sort and another part is real delete. 

And for delete, we should add a TODO for auditing in the future

Reply:
👌